client-websocket/src/main/java/com/github/twitch4j/client/websocket/WebsocketConnection.java

```diff
@@ -0,0 +1,237 @@
+            public void onDisconnected(WebSocket websocket, WebSocketFrame serverCloseFrame, WebSocketFrame clientCloseFrame, boolean closedByServer) {
+                if (!connectionState.equals(WebsocketConnectionState.DISCONNECTING)) {
+                    log.info("Connection to WebSocket [{}] lost! Retrying soon ...", config.baseUrl());
+
+                    // connection lost - reconnecting
+                    if (backoffClearer != null) backoffClearer.cancel(false);
+                    config.taskExecutor().schedule(() -> reconnect(), config.backoffStrategy().get(), TimeUnit.MILLISECONDS);
+                } else {
+                    connectionState = WebsocketConnectionState.DISCONNECTED;
+                    log.info("Disconnected from WebSocket [{}]!", config.baseUrl());
+                }
+            }
+
+            @Override
+            public void onFrameSent(WebSocket websocket, WebSocketFrame frame) {
+                if (frame != null && frame.isPingFrame()) {
+                    lastPing.compareAndSet(0L, System.currentTimeMillis());
+                }
+            }
+
+            @Override
+            public void onPongFrame(WebSocket websocket, WebSocketFrame frame) {
+                final long last = lastPing.getAndSet(0L);
+                if (last > 0) {
+                    latency = System.currentTimeMillis() - last;
+                    log.trace("TwitchChat: Round-trip socket latency recorded at {} ms.", latency);
+                }
+            }
+        };
+    }
+
+    /**
+     * Connect to the WebSocket
+     */
+    @Synchronized
+    public void connect() {
+        if (connectionState.equals(WebsocketConnectionState.DISCONNECTED) || connectionState.equals(WebsocketConnectionState.RECONNECTING)) {
+            try {
+                // hook: on pre connect
+                if (config.onPreConnect() != null)
+                    config.onPreConnect().run();
+
+                // Change Connection State
+                connectionState = WebsocketConnectionState.CONNECTING;
+
+                // init websocket
+                webSocket = webSocketFactory.createSocket(config.baseUrl());
+                webSocket.setPingInterval(config.wsPingPeriod());
+                webSocket.clearListeners();
+                webSocket.addListener(webSocketAdapter);
+
+                // connect
+                this.webSocket.connect();
+
+                // hook: post connect
+                if (config.onPostConnect() != null)
+                    config.onPostConnect().run();
+            } catch (Exception ex) {
+                log.error("connection to webSocket server {} failed: retrying ...", config.baseUrl(), ex);
+                // Sleep before trying to reconnect
+                try {
+                    config.backoffStrategy().sleep();
+                } catch (Exception ignored) {
+
+                } finally {
+                    // reconnect
+                    reconnect();
```

Comment:
Can we adding some connection retry count (in configuration) when should not do any force connection in `n` retries? In default `-1` when will be a loop connection to retrying until killing machine?

Reply:
delays between connection attempts are currently handled by the backoffStrategy a few lines higher `config.backoffStrategy().sleep();`

https://github.com/twitch4j/twitch4j/blob/feature/websocket-common/client-websocket/src/main/java/com/github/twitch4j/client/websocket/WebsocketConnectionConfig.java#L49-L58

Reply:
didn't add about it stops after `n` count retries.

Reply:
`ExponentialBackoffStrategy` actually has a `maxRetries` field that can be configured. however's it's not currently honored by the socket; if `config.backoffStrategy().sleep()` returns false, we should not attempt reconnect

Reply:
Fixed bug in https://github.com/twitch4j/twitch4j/pull/552/commits/bc4bfea28a400b1c6736943fe9cacdbe49fb4607 - though I don't think users want `n` to be non-negative by default (so it will retry forever, unless user configures otherwise)